### ISSUE TITLE
Inconsistent calculation of Replycount in Threads table for threads created prior to MM 5.29.0.

### ISSUE THREAD

**Issue by someone-somenet-org** [mattermost/mattermost-server#16321]
While trying to understand the db changes that came with MM 5.29.0 in order to update my [db+fs-cleaner](https://git.somenet.org/pub/jan/mattermost-privileged.git/blob/HEAD:/maintenance/db.py) I came across what looks like inconsistent behavior in the `Threads` table.

The `Replycount` column in the `Threads` table seems to be inconsistently calculated for existing threads before updating to MM 5.29.0.

For new threads `Threads.Replycount` seems to only include not deleted posts. When a post is deleted in a thread, the thread's `Replycount` is decremented. So it seems to me that this is the intended behavior.

Now when adding a reply to any thread that existed prior to updating to 5.29.0 the entry added to the `Threads` table will also count posts that were deleted (= post's deteleat > 0).
It also wont fix itself when posting more in pre-5.29.0 threads.

#### Steps to reproduce
+ Install MM 5.28.x
+ Create thread with 5 posts. (OP-post + 4 replies)
+ Delete one reply. (Thread has 1+3 visible posts now)
+ Update to 5.29.0
+ Add a post to that thread. (Thread has 1+4 visible posts now)
+ Look at `Threads` table. `Replycount = 5`.

#### Expected behavior
When doing the exact same steps in 5.29.0 you end up with `Threads.Replycount = 4` - the deleted posts are excluded from the postcount.

#### Observed behavior (that appears unintentional) 
When replying to threads that were created before the update to 5.29.0 they will get an entry in the Threads table, including the deleted posts (= post's deteleat > 0) in the `Replycount` column.

#### Possible fixes
It is obviously really early in development of the new threads (I cant await them!)
But it looks like inconsistent behavior to me, so better report it, before it causes some issues later.

**Comment by reflog** [comment]
Hi @someone-somenet-org , thanks for reporting this. I tried reproducing using a small test, but was unable to. Here's the test fir 5.29 , A thread with 1+4 posts, deleting one reply, then counting replies returns 3.
Did I miss something in the test?
```
		Client := th.Client

		rpost, resp := Client.CreatePost(&model.Post{ChannelId: th.BasicChannel.Id, Message: "testMsg"})
		CheckNoError(t, resp)
		CheckCreatedStatus(t, resp)
		Client.CreatePost(&model.Post{RootId: rpost.Id, ChannelId: th.BasicChannel.Id, Message: "testMsg"})
		Client.CreatePost(&model.Post{RootId: rpost.Id, ChannelId: th.BasicChannel.Id, Message: "testMsg"})
		a, _ := Client.CreatePost(&model.Post{RootId: rpost.Id, ChannelId: th.BasicChannel.Id, Message: "testMsg"})
		Client.CreatePost(&model.Post{RootId: rpost.Id, ChannelId: th.BasicChannel.Id, Message: "testMsg"})

		defer th.App.Srv().Store.Post().PermanentDeleteByUser(th.BasicUser.Id)
		Client.DeletePost(a.Id)
		uss, resp := th.Client.GetUserThreads(th.BasicUser.Id, model.GetUserThreadsOpts{
			Page:     0,
			PageSize: 30,
		})
		require.Nil(t, resp.Error)
		require.EqualValues(t, uss.Threads[0].ReplyCount, 3)
```

**Comment by someone-somenet-org** [comment]
Its a Migration issue. You need to start a thread and delete a post in a version before MM 5.29.0 and then update to MM 5.29.0 and then post once to trigger calculation. 
(Or just create the thread, delete one post, `truncate table threads`, invalidate mm Cache, add a post to the thread)

Im not sure - on mobile only currently - but I would say https://github.com/mattermost/mattermost-server/blob/b3aae7406c213018a5a5a8c3f7797948522e92c0/store/sqlstore/post_store.go#L1982 is missing a `and deleteat = 0`

**Comment by reflog** [comment]
@someone-somenet-org great catch, thank you!

**Comment by reflog** [comment]
Tracking https://mattermost.atlassian.net/browse/MM-30917

**Comment by amyblais** [comment]
Closing as this was fixed in v5.29.1,
